activation/handler.go

```diff
@@ -281,56 +277,22 @@ func (h *Handler) handleAtx(
 		return nil, fmt.Errorf("%w: atx want %s, got %s", errWrongHash, expHash.ShortString(), id.ShortString())
 	}
 
-	// Check if processing is already in progress
-	h.inProgressMu.Lock()
-	if sub, ok := h.inProgress[id]; ok {
-		ch := make(chan error, 1)
-		h.inProgress[id] = append(sub, ch)
-		h.inProgressMu.Unlock()
-		h.logger.Debug("atx is already being processed. waiting for result",
-			log.ZContext(ctx),
-			zap.Stringer("atx_id", id),
-		)
-		select {
-		case err := <-ch:
-			h.logger.Debug("atx processed in other task",
-				log.ZContext(ctx),
-				zap.Stringer("atx_id", id),
-				zap.Error(err),
-			)
-			return nil, err
-		case <-ctx.Done():
-			return nil, ctx.Err()
+	key := string(id.Bytes())
+	proof, err, _ := h.inProgress.Do(key, func() (any, error) {
+		h.logger.Info("handling incoming atx", log.ZContext(ctx), zap.Stringer("atx_id", id), zap.Int("size", len(msg)))
+
+		switch atx := opaqueAtx.(type) {
+		case *wire.ActivationTxV1:
+			return h.v1.processATX(ctx, peer, atx, msg, receivedTime)
+		case *wire.ActivationTxV2:
+			return h.v2.processATX(ctx, peer, atx, msg, receivedTime)
+		default:
+			panic("unreachable")
 		}
-	}
+	})
+	h.inProgress.Forget(key)
 
-	h.inProgress[id] = []chan error{}
-	h.inProgressMu.Unlock()
-	h.logger.Info("handling incoming atx",
-		log.ZContext(ctx),
-		zap.Stringer("atx_id", id),
-		zap.Int("size", len(msg)),
-	)
-
-	var proof *mwire.MalfeasanceProof
-
-	switch atx := opaqueAtx.(type) {
-	case *wire.ActivationTxV1:
-		proof, err = h.v1.processATX(ctx, peer, atx, msg, receivedTime)
-	case *wire.ActivationTxV2:
-		proof, err = h.v2.processATX(ctx, peer, atx, msg, receivedTime)
-	default:
-		panic("unreachable")
-	}
-
-	h.inProgressMu.Lock()
-	defer h.inProgressMu.Unlock()
-	for _, ch := range h.inProgress[id] {
-		ch <- err
-		close(ch)
-	}
-	delete(h.inProgress, id)
-	return proof, err
+	return proof.(*mwire.MalfeasanceProof), err
```

Comment:
won't this `panic` if `proof` is `nil`?

Reply:
https://go.dev/play/p/kAWFunhwsba

Reply:
The tests cover this scenario. It won't panic because here, `proof` is not dereferenced, but cast. The result is `nil` of different pointer type. 

>[go.dev/play/p/kAWFunhwsba](https://go.dev/play/p/kAWFunhwsba)

Your example panics not because `vv` is nil, but because its type is not `*TT`. Check out this playground: https://go.dev/play/p/ZHZO9BVzJfE.